matrix_sdk/src/room/joined.rs

```diff
@@ -0,0 +1,520 @@
+            reason
+        });
+        self.client.send(request, None).await?;
+        Ok(())
+    }
+
+    /// Kick a user out of this room.
+    ///
+    /// # Arguments
+    ///
+    /// * `user_id` - The `UserId` of the user that should be kicked out of the room.
+    ///
+    /// * `reason` - Optional reason why the room member is being kicked out.
+    pub async fn kick_user(&self, user_id: &UserId, reason: Option<&str>) -> Result<()> {
+        let request = assign!(kick_user::Request::new(self.inner.room_id(), user_id), {
+            reason
+        });
+        self.client.send(request, None).await?;
+        Ok(())
+    }
+
+    /// Invite the specified user by `UserId` to this room.
+    ///
+    /// # Arguments
+    ///
+    /// * `user_id` - The `UserId` of the user to invite to the room.
+    pub async fn invite_user_by_id(&self, user_id: &UserId) -> Result<()> {
+        let recipient = InvitationRecipient::UserId { user_id };
+
+        let request = invite_user::Request::new(self.inner.room_id(), recipient);
+        self.client.send(request, None).await?;
+
+        Ok(())
+    }
+
+    /// Invite the specified user by third party id to this room.
+    ///
+    /// # Arguments
+    ///
+    /// * `invite_id` - A third party id of a user to invite to the room.
+    pub async fn invite_user_by_3pid(&self, invite_id: Invite3pid<'_>) -> Result<()> {
+        let recipient = InvitationRecipient::ThirdPartyId(invite_id);
+        let request = invite_user::Request::new(self.inner.room_id(), recipient);
+        self.client.send(request, None).await?;
+
+        Ok(())
+    }
+
+    /// Send a request to notify this room of a user typing.
+    ///
+    /// # Arguments
+    ///
+    /// * `typing` - Whether the user is typing, and how long.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use std::time::Duration;
+    /// # use matrix_sdk::{
+    /// #     Client, SyncSettings,
+    /// #     api::r0::typing::create_typing_event::Typing,
+    /// #     identifiers::room_id,
+    /// # };
+    /// # use futures::executor::block_on;
+    /// # use url::Url;
+    /// # block_on(async {
+    /// # let homeserver = Url::parse("http://localhost:8080").unwrap();
+    /// # let mut client = Client::new(homeserver).unwrap();
+    /// # let room_id = room_id!("!test:localhost");
+    /// # let room = client
+    /// #    .get_joined_room(&room_id!("!SVkFJHzfwvuaIEawgC:localhost"))
+    /// #    .unwrap();
+    /// # room
+    ///     .typing_notice(Typing::Yes(Duration::from_secs(4)))
+    ///     .await
+    ///     .expect("Can't get devices from server");
+    /// # });
+    ///
+    /// ```
+    pub async fn typing_notice(&self, typing: impl Into<Typing>) -> Result<()> {
+        let user_id = self
+            .client
+            .user_id()
+            .await
+            .ok_or(Error::AuthenticationRequired)?;
+        let request = TypingRequest::new(&user_id, self.inner.room_id(), typing.into());
```

Comment:
Please add a TODO here that we should deduplicate the request for users. E.g. do nothing if a typing notice is being sent or is already active.

Reply:
I think a user can stop a typing notice before the timeout it's over so the user may resend the same request. Also i think the user can update the timeout will a typing notice is still valid.

Reply:
Yes, stopping should be allowed, and sending out another one when the notice is expired or is about to expire should be possible as well.

We still want to hinder people in sending out 10 requests in quick succession, essentially sending a request every time a user types a letter. I understand that every client could deduplicate this themselves, chances are that they will mess it up. This is mainly motivated by Fractal sending out multiple typing notices until it finally gets an answer for one of the requests, and then it correctly stops spamming the server.

Reply:
> This is mainly motivated by Fractal sending out multiple typing notices until it finally gets an answer for one of the requests, and then it correctly stops spamming the server.
:rofl:  I didn't know Fractal was doing this and it should be fixed there.

Yes, stopping should be allowed, and sending out another one when the notice is expired or is about to expire should be possible as well.
Ok i will do this then.

Reply:
We can just copy the implementation from [weechat-matrix](https://github.com/poljar/weechat-matrix-rs/blob/master/src/room/mod.rs#L686). While Fractal may be fixed, it's likely that other SDK consumers will fall in the same trap, so I think it's wise to push the functionality into the SDK.

Reply:
i think https://github.com/matrix-org/matrix-rust-sdk/pull/172/commits/4ab1069d9c860e0d397a9da531ccce9d10a8ae6a should to what you want. It allows only to sent requests if there isn't an active typing notice or the typing notice state changes between Yes and No.